### PR TITLE
pkcs11 symmetric ciphers

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Depends on https://github.com/OP-TEE/optee_os/pull/4023

**Comment by vesajaaskelainen** [comment]
There seems to be missing C_GenerateKey(). Or is this planned for future PRs?.

Eg. Create AES secret key within the OP-TEE.

File: libckteec/src/pkcs11_processing.c [hunk 1]
```diff
@@ -15,6 +15,33 @@
...
+{
+	TEEC_SharedMemory *shm = NULL;
+
+	if (!in)
+		return NULL;
+
+	shm = ckteec_register_shm(in, len, CKTEEC_SHM_IN);
+	if (!shm) {
+		/*
+		 * Input buffer cannot be registered: allocated a
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/pkcs11_processing.c:33]: I'll reword it slightly.

File: libckteec/src/pkcs11_processing.c [hunk 2]
```diff
@@ -15,6 +15,33 @@
...
+		 * temporary input shared buffer rather than failing.
+		 */
+		shm = ckteec_alloc_shm(len, CKTEEC_SHM_IN);
```
Comment (jenswi-linaro) [inline libckteec/src/pkcs11_processing.c:36]: Question for @etienne-lms: is this a likely scenario where registration fails, but a memory allocation will succeed?

File: libckteec/src/pkcs11_processing.c [hunk 3]
```diff
@@ -106,3 +133,240 @@ CK_RV ck_destroy_object(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE obj)
...
 }
+
+CK_RV ck_encdecrypt_init(CK_SESSION_HANDLE session,
+			 CK_MECHANISM_PTR mechanism,
+			 CK_OBJECT_HANDLE key,
+			 int decrypt)
+{
+	CK_RV rv = CKR_GENERAL_ERROR;
+	TEEC_SharedMemory *ctrl = NULL;
+	struct serializer obj;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/pkcs11_processing.c:144]: Thanks, I'll fix

File: libckteec/src/pkcs11_processing.c [hunk 4]
```diff
@@ -106,3 +133,240 @@ CK_RV ck_destroy_object(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE obj)
...
+	buf += sizeof(key_handle);
+
+	memcpy(buf, obj.buffer, obj.size);
+
+	rv = ckteec_invoke_ctrl(decrypt ? PKCS11_CMD_DECRYPT_INIT :
+				PKCS11_CMD_ENCRYPT_INIT, ctrl);
+
+bail:
+	release_serial_object(&obj);
+	ckteec_free_shm(ctrl);
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/pkcs11_processing.c:185]: I'm not sure that matters much. What would be the gain?

File: libckteec/src/serialize_ck.c [hunk 5]
```diff
@@ -192,3 +192,120 @@ CK_RV serialize_ck_attributes(struct serializer *obj,
...
+ * Serialization of CK mechanism parameters
+ *
+ * Most mechanism have no parameters.
+ * Some mechanism have a single 32bit parameter.
+ * Some mechanism have a specific parameter structure which may contain
+ * indirected data (data referred by a buffer pointer).
+ *
+ * Below are each structure specific mechanisms parameters.
+ *
+ * Be careful that CK_ULONG based types translate to 32bit ulong fields in
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serialize_ck.c:206]: Agree, I'll drop the comment.

File: libckteec/src/serialize_ck.c [hunk 6]
```diff
@@ -192,3 +192,120 @@ CK_RV serialize_ck_attributes(struct serializer *obj,
...
+ * the PKCS11 ABI.
+ */
+
+/*
+ * typedef struct CK_AES_CTR_PARAMS {
+ *	CK_ULONG ulCounterBits;
+ *	CK_BYTE cb[16];
+ * } CK_AES_CTR_PARAMS;
+ */
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serialize_ck.c:215]: Agree, I'll drop this.

**Comment by jenswi-linaro** [comment]
> There seems to be missing C_GenerateKey(). Or is this planned for future PRs?.
> 
> Eg. Create AES secret key within the OP-TEE.

Yes, we'll take that later. I'm trying to introduce as small bits as possible to make the reviews easier.

File: libckteec/src/pkcs11_processing.c (hunk 2 above)
Comment (etienne-lms) [reply libckteec/src/pkcs11_processing.c:37]: From the inline comment, it looks like there were issues with registering read-only memory that got written to at some point. I will check if this is still accurate.

File: libckteec/src/pkcs11_processing.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libckteec/src/pkcs11_processing.c:186]: True, `obj` and `ctrl` are not related so no importance here.

File: libckteec/src/pkcs11_processing.c (hunk 2 above)
Comment (jenswi-linaro) [reply libckteec/src/pkcs11_processing.c:37]: @etienne-lms, any news on this?

File: libckteec/src/pkcs11_processing.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/pkcs11_processing.c:186]: OK, I'll swap the lines

File: libckteec/src/pkcs11_processing.c [hunk 7]
```diff
@@ -106,3 +133,240 @@ CK_RV ck_destroy_object(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE obj)
...
+
+	if (out_len && (rv == CKR_OK || rv == CKR_BUFFER_TOO_SMALL))
+		*out_len = out_size;
+
+	if (rv == CKR_BUFFER_TOO_SMALL && out_size && !out)
+		rv = CKR_OK;
+
+bail:
+	ckteec_free_shm(in_shm);
+	ckteec_free_shm(out_shm);
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/pkcs11_processing.c:317]: OK

**Comment by jenswi-linaro** [comment]
All comments addressed. I'll wait with squashing until @etienne-lms has answered my question above.

File: libckteec/src/pkcs11_processing.c (hunk 2 above)
Comment (etienne-lms) [reply libckteec/src/pkcs11_processing.c:37]: sorry, I've started but switched to something else a forgot this. I'll look back.

If I remember well, there were some `const ` variable in the xtest, likely for pkcs11 object attributes that caused attributes.

Reading back this comment today, I think the registering should not fail. As it sounds weird to expect it to fail, I think we can remove this 2nd try and return an error. I we found it's needed, I'll see how to address.

Comment (jenswi-linaro) [reply libckteec/src/pkcs11_processing.c:37]: OK, thanks. I'll drop this then.

**Comment by jenswi-linaro** [comment]
Commits squashed and tag applied.

**Comment by vesajaaskelainen** [comment]
Went thru it once again and didn't spot anything.

**Pull request merged**